// @wordpress/components/build-module/combobox-control/index.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { useCombobox } from 'downshift';
import classnames from 'classnames';
/**
 * Internal dependencies
 */

import { Button, Dashicon } from '../';

var itemToString = function itemToString(item) {
  return item && item.name;
};

export default function ComboboxControl(_ref) {
  var className = _ref.className,
      hideLabelFromVision = _ref.hideLabelFromVision,
      label = _ref.label,
      items = _ref.options,
      onInputValueChange = _ref.onInputValueChange,
      onSelectedItemChange = _ref.onChange,
      _selectedItem = _ref.value;

  var _useCombobox = useCombobox({
    initialSelectedItem: items[0],
    items: items,
    itemToString: itemToString,
    onInputValueChange: onInputValueChange,
    onSelectedItemChange: onSelectedItemChange,
    selectedItem: _selectedItem
  }),
      getLabelProps = _useCombobox.getLabelProps,
      getToggleButtonProps = _useCombobox.getToggleButtonProps,
      getComboboxProps = _useCombobox.getComboboxProps,
      getInputProps = _useCombobox.getInputProps,
      getMenuProps = _useCombobox.getMenuProps,
      getItemProps = _useCombobox.getItemProps,
      isOpen = _useCombobox.isOpen,
      highlightedIndex = _useCombobox.highlightedIndex,
      selectedItem = _useCombobox.selectedItem;

  var menuProps = getMenuProps({
    className: 'components-combobox-control__menu'
  }); // We need this here, because the null active descendant is not
  // fully ARIA compliant.

  if (menuProps['aria-activedescendant'] && menuProps['aria-activedescendant'].slice(0, 'downshift-null'.length) === 'downshift-null') {
    delete menuProps['aria-activedescendant'];
  }

  return createElement("div", {
    className: classnames('components-combobox-control', className)
  }, createElement("label", getLabelProps({
    className: classnames('components-combobox-control__label', {
      'screen-reader-text': hideLabelFromVision
    })
  }), label), createElement("div", getComboboxProps({
    className: 'components-combobox-control__button'
  }), createElement("input", getInputProps({
    className: 'components-combobox-control__button-input'
  })), createElement(Button, getToggleButtonProps({
    // This is needed because some speech recognition software don't support `aria-labelledby`.
    'aria-label': label,
    'aria-labelledby': undefined,
    className: 'components-combobox-control__button-button'
  }), createElement(Dashicon, {
    icon: "arrow-down-alt2",
    className: "components-combobox-control__button-icon"
  }))), createElement("ul", menuProps, isOpen && items.map(function (item, index) {
    return (// eslint-disable-next-line react/jsx-key
      createElement("li", getItemProps({
        item: item,
        index: index,
        key: item.key,
        className: classnames('components-combobox-control__item', {
          'is-highlighted': index === highlightedIndex
        }),
        style: item.style
      }), item === selectedItem && createElement(Dashicon, {
        icon: "saved",
        className: "components-combobox-control__item-icon"
      }), item.name)
    );
  })));
}
