// reakit/es/Grid/GridRow.js
import { a as _objectSpread2 } from '../_rollupPluginBabelHelpers-1f0bf8c2.js';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import 'reakit-utils/shallowEqual';
import '../Box/Box.js';
import 'react';
import 'reakit-utils/useForkRef';
import 'reakit-utils/useIsomorphicEffect';
import '../findEnabledItemById-8ddca752.js';
import '../__keys-6b7fda7f.js';
import '../Group/Group.js';
import '../Id/IdProvider.js';
import '../Id/Id.js';
import { useCompositeGroup } from '../Composite/CompositeGroup.js';
import { b as GRID_ROW_KEYS } from '../__keys-49e65cac.js';

var unstable_useGridRow = createHook({
  name: "GridRow",
  compose: useCompositeGroup,
  keys: GRID_ROW_KEYS,
  useProps: function useProps(_, htmlProps) {
    return _objectSpread2({
      role: "row"
    }, htmlProps);
  }
});
var unstable_GridRow = createComponent({
  as: "div",
  useHook: unstable_useGridRow
});

export { unstable_GridRow, unstable_useGridRow };
