// reakit/es/Grid/GridCell.js
import { a as _objectSpread2 } from '../_rollupPluginBabelHelpers-1f0bf8c2.js';
import { createComponent } from 'reakit-system/createComponent';
import { createHook } from 'reakit-system/createHook';
import 'reakit-utils/shallowEqual';
import '../Box/Box.js';
import 'react';
import 'reakit-utils/useForkRef';
import 'reakit-utils/isButton';
import 'reakit-warning';
import 'reakit-utils/useLiveRef';
import 'reakit-utils/isSelfTarget';
import 'reakit-utils/useIsomorphicEffect';
import 'reakit-utils/hasFocusWithin';
import 'reakit-utils/isPortalEvent';
import 'reakit-utils/getActiveElement';
import 'reakit-utils/tabbable';
import '../Tabbable/Tabbable.js';
import '../Clickable/Clickable.js';
import 'reakit-utils/createOnKeyDown';
import 'reakit-utils/getDocument';
import '../getCurrentId-5aa9849e.js';
import '../__keys-6b7fda7f.js';
import '../Id/IdProvider.js';
import '../Id/Id.js';
import 'reakit-utils/isTextField';
import 'reakit-utils/ensureFocus';
import '../setTextFieldValue-78ab5d56.js';
import { useCompositeItem } from '../Composite/CompositeItem.js';
import { a as GRID_CELL_KEYS } from '../__keys-49e65cac.js';

var unstable_useGridCell = createHook({
  name: "GridCell",
  compose: useCompositeItem,
  keys: GRID_CELL_KEYS,
  useProps: function useProps(_, htmlProps) {
    return _objectSpread2({
      role: "gridcell"
    }, htmlProps);
  }
});
var unstable_GridCell = createComponent({
  as: "span",
  memo: true,
  useHook: unstable_useGridCell
});

export { unstable_GridCell, unstable_useGridCell };
