// reakit/es/Popover/index.js
import '../_rollupPluginBabelHelpers-1f0bf8c2.js';
import 'reakit-system/createComponent';
import 'reakit-system/createHook';
import 'reakit-utils/shallowEqual';
import '../Box/Box.js';
import 'react';
import 'reakit-utils/useForkRef';
import 'reakit-utils/isButton';
import 'reakit-warning';
import 'reakit-utils/useLiveRef';
import 'reakit-utils/isSelfTarget';
import 'reakit-utils/useIsomorphicEffect';
import 'reakit-utils/hasFocusWithin';
import 'reakit-utils/isPortalEvent';
import 'reakit-utils/getActiveElement';
import 'reakit-utils/tabbable';
import '../Tabbable/Tabbable.js';
import '../Clickable/Clickable.js';
import '../Button/Button.js';
import 'reakit-utils/useSealedState';
import 'reakit-system/useCreateElement';
import 'reakit-utils/getDocument';
import 'reakit-utils/getNextActiveElementOnBlur';
import '../Id/IdProvider.js';
import 'reakit-utils/ensureFocus';
import '../Id/IdState.js';
import '../__keys-e6a5cfbe.js';
import '../Disclosure/DisclosureContent.js';
import 'react-dom';
import '../Portal/Portal.js';
import 'reakit-utils/removeItemFromArray';
import '../MenuContext-6af6cf92.js';
import '../Dialog/Dialog.js';
import 'body-scroll-lock';
import 'reakit-utils/useUpdateEffect';
import 'reakit-utils/closest';
import 'reakit-utils/contains';
import '../DialogBackdropContext-8775f78b.js';
import 'reakit-utils/isEmpty';
import '../__keys-ed7b48af.js';
import '../Dialog/DialogBackdrop.js';
import '../Disclosure/Disclosure.js';
import '../Dialog/DialogDisclosure.js';
import '../Disclosure/DisclosureState.js';
import '../Dialog/DialogState.js';
import '../__keys-26bb1730.js';
export { Popover, usePopover } from './Popover.js';
export { PopoverArrow, usePopoverArrow } from './PopoverArrow.js';
export { PopoverDisclosure, usePopoverDisclosure } from './PopoverDisclosure.js';
import '@popperjs/core';
export { usePopoverState } from './PopoverState.js';
export { PopoverBackdrop, usePopoverBackdrop } from './PopoverBackdrop.js';
